[CORE-11981] pandaproxy/build: split dependencies

BenPope opened the pull request:
Split pandaproxy into several targets.

## Backports Required

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes

* none

BenPope committed:
pandaproxy: Split dependencies: Phase 1

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
pandaproxy: Split dependencies: Phase 2

Move pandaproxy:rest_server to pandaproxy/rest:server.

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
pandaproxy: Split dependencies: Phase 3

Move pandaproxy:schema_registry* to pandaproxy/schema_registry*

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope commented on src/v/pandaproxy/schema_registry/BUILD:
```diff
@@ -0,0 +1,185 @@
+        "avro.h",
+        "compatibility.h",
+        "configuration.h",
+        "error.h",
+        "errors.h",
+        "exceptions.h",
+        "fwd.h",
```
It might be odd that I've included these headers, when they also appear in `:core`, but it saves doing:
```bazel
    deps = [
        "//src/v/pandaproxy:schema_registry:core",
        "//src/v/pandaproxy:schema_registry:server",
    ],
```

michael-redpanda replied:
You could have a `fwd` package that includes just `fwd.h` - there's some examples of that in the code base where the public deps include it but the implementation_deps could include the defined structures that were declared in `fwd`

michael-redpanda replied:
I'm not sure exactly what's idiomatic for bazel, but I suspect having the same header in multiple... packages?.... is not

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--c0b5a643-3cf1-47f5-9150-7c7479595b48-->
<!--TEST RESULTS-->
<!--79bcabc8-cdc2-4878-8105-4b87887aa42c-->
<!--TEST RESULTS-->
<!--7e59ba10-142f-4a93-aa01-3111df583199-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/66793">build#66793</a></summary>


| test_class                 | test_method                               | test_arguments                                         | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                     |
|:---------------------------|:------------------------------------------|:-------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------|
| raft_fixture               | leadership_transfer_delay                 |                                                        | unit        | https://buildkite.com/redpanda/redpanda/builds/66793#0197381d-ed59-4927-b076-087275c451b5 | FAIL          | 0/1      |                                                                                                                                                            |
| ArchivalTest               | test_single_partition_leadership_transfer | {"cloud_storage_type": 2}                              | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66793#019738cb-8edb-4086-9e06-e7211c893fff | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS  |
| ConsumerGroupTest          | test_consumer_is_removed_when_timedout    | {"static_members": false}                              | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66793#019738cb-8edb-4086-9e06-e7211c893fff | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS  |
| DatalakeClusterRestoreTest | test_slow_tiered_storage_dupe_records     | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66793#019738cb-8edb-4086-9e06-e7211c893fff | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS |
| PartitionStateAPItest      | test_local_summary                        |                                                        | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66793#019738cb-8edb-4086-9e06-e7211c893fff | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS  |
| TopicRecoveryTest          | test_fast3                                | {"cloud_storage_type": 2}                              | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66793#019738cb-8edb-4086-9e06-e7211c893fff | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS  |
| TransactionsStreamsTest    | consumer_offsets_retention_test           |                                                        | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66793#019738cb-8edb-4086-9e06-e7211c893fff | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS  |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/66846">build#66846</a></summary>


| test_class                | test_method                           | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:--------------------------|:--------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ClusterHealthOverviewTest | cluster_health_overview_baseline_test |                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66846#01973c10-ff12-4b77-afdc-1fa509bcf4f7 | FLAKY         | 20/21    | upstream reliability is '97.07602339181285'. current run reliability is '95.23809523809523'. drift is 1.83793 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/66987">build#66987</a></summary>


| test_class               | test_method          | test_arguments                                                                                                                       | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                |
|:-------------------------|:---------------------|:-------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| RandomNodeOperationsTest | test_node_operations | {"cloud_storage_type": 2, "enable_failures": true, "mixed_versions": false, "with_chunked_compaction": false, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66987#019745b7-c273-444a-880e-780421914a3a | FLAKY         | 17/21    | upstream reliability is '100.0'. current run reliability is '80.95238095238095'. drift is 19.04762 and the allowed drift is set to 50. The test should PASS           |
| DisablingPartitionsTest  | test_disable         |                                                                                                                                      | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66987#019745ba-dcf3-4c5b-9056-27b6bedf5654 | FLAKY         | 17/21    | upstream reliability is '88.2494004796163'. current run reliability is '80.95238095238095'. drift is 7.29702 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

pgellert commented on src/v/pandaproxy/BUILD:
```diff
@@ -13,207 +25,198 @@ seastar_cc_swagger_library(
+        ":json",
+        ":parsing",
+        "//src/v/base",
+        "//src/v/kafka/protocol",
```
nit: should some of these be `implementation_deps` instead of `deps`?

pgellert commented on src/v/pandaproxy/schema_registry/BUILD:
```diff
@@ -0,0 +1,186 @@
+        "service.cc",
+        "sharded_store.cc",
+        "validation.cc",
+    ],
```
Would there be a way to split server such most modules outside of `src/v/pandaproxy` only have to depend on things outside of the server+handlers implementation? That way we could avoid recompiling most things when making changes in the handlers or the server.

(nit: This part of SR, ie. things without the http handlers is what I would call core, and maybe the existing core could be called types.)

BenPope replied:
> Would there be a way to split server such most modules outside of src/v/pandaproxy only have to depend on things outside of the server+handlers implementation?

I couldn't get it to work, unfortunately.

BenPope replied:
OK, done.

pgellert reviewed (commented):
🔥

BenPope commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/c8dd882fbda2dad9db6e8366597abbc6a0888d05..bc7daa5c32d290052d383e4fee15c7ce47fc1b87)
* Renamed `core` to `types`
* Renamed `server` to `core`
* `server` now exposes fewer headers
* Introduced `registry` for subsystems that need more access than `server`
* Moved many `deps` to `implementation_deps`

BenPope commented:
Changes in [force-push(https://github.com/redpanda-data/redpanda/compare/cc625ce6e18486862fcda462a283a7c55d4a9557..d16c5919f4fa3538fb3ff7e01ef17b97dac4e675)
* Move some stuff out of `core` and into `server`

michael-redpanda commented on src/v/pandaproxy/BUILD:
```diff
@@ -5,6 +5,19 @@ load("//src/v/schema/protobuf:well_known.bzl", "well_known_cc_protos")
 seastar_cc_swagger_library(
     name = "rest_swagger",
     src = "api/api-doc/rest.json",
+    visibility = ["//visibility:public"],
```
nit: probably doesn't have to be public?

IoannisRP commented on src/v/pandaproxy/schema_registry/BUILD:
```diff
@@ -0,0 +1,263 @@
+    ],
+)
+
+redpanda_cc_library(
+    name = "registry",
+    srcs = [],
+    hdrs = [
+        "api.h",
+        "avro.h",
+        "compatibility.h",
+        "configuration.h",
+        "fwd.h",
+        "json.h",
+        "protobuf.h",
+        "schema_getter.h",
+        "seq_writer.h",
+        "service.h",
+        "sharded_store.h",
+        "types.h",
+        "validation.h",
+    ],
+    implementation_deps = [
+        "//src/v/cluster",
+    ],
+    include_prefix = "pandaproxy/schema_registry",
+    visibility = ["//visibility:public"],
+    deps = [
+        ":server",
+        "//src/v/pandaproxy:core",
+    ],
+)
```
I assume this set of headers is redefined as a library to be used out of schema registry, to limit what headers the external modules can see, right? Or is it something else?

Either way, it feels wrong to have the headers duplicated. Would it be easy to move the `cc` together with their headers in a single library? This would be instead of having both `h` and `cc` as part of `core` and the `h` as a header-only lib. Or are the cyclical dependencies issue?

IoannisRP reviewed (commented):
Very nice modularization of the pandaproxy! :fire:

dotnwat commented:
nice!
